pkg/kv/kvprober/kvprober.go

```diff
@@ -306,11 +308,42 @@ func (p *Prober) Metrics() Metrics {
 	return p.metrics
 }
 
+// LogLeaseholderInfo searches through the given tracing recording for log messages
+// indicating updated range information, extracts range IDs and leaseholder node IDs,
+// and logs this information only if the leaseholder for a range has been updated.
+func (p *Prober) LogLeaseholderInfo(ctx context.Context, recording tracingpb.Recording) {
+	// Find the log message containing information about updated ranges
+	informationLog, isValid := recording.FindLogMessage("received updated range")
+
+	if isValid {
+		// Regular expressions to extract range IDs and leaseholder node IDs
+		rangeRegex := regexp.MustCompile(`desc: r(\d+)`)
+		leaseRegex := regexp.MustCompile(`repl=\(n(\d+)`)
+
+		// Extract all occurrences of range IDs and leaseholder node IDs from the log message
+		ranges := rangeRegex.FindAllStringSubmatch(informationLog, -1)
+		leaseholders := leaseRegex.FindAllStringSubmatch(informationLog, -1)
+
+		for i := 0; i < len(ranges); i++ {
+			// Create a new context with tags for the current range ID and leaseholder node ID
+			newCtx := logtags.AddTag(ctx, "range", ranges[i][1])
+			newCtx = logtags.AddTag(newCtx, "leaseholder", leaseholders[i][1])
+
+			// Log the range and leaseholder information using the new context
+			log.Health.Infof(newCtx, "range %s has leaseholder %s", ranges[i][1], leaseholders[i][1])
```

Comment:
I think instead of logging via a separate log line, we should include the leaseholder in existing logs lines such as what is below. So perhaps this function should return the leaseholder. Can be called from `readProbeImpl`, etc.

```
	log.Health.Infof(ctx, "kv.Get(%s), r=%v returned success in %v", step.Key, step.RangeID, d)
```

This will make it possible to do Datadog queries, where we break down the error rate by leaseholder. Similar to: https://us5.datadoghq.com/logs?query=service%3Acockroachdb%20%40friendly_id%3Acrl-prod-n89%20status%3Aerror%20%22unable%20to%20connect%20%28is%20the%20peer%20up%20and%20reachable%3F%29%22%20&agg_m=count&agg_m_source=base&agg_q=%40tags.n&agg_q_source=base&agg_t=count&analyticsOptions=%5B%22bars%22%2C%22dog_classic%22%2Cnull%2Cnull%5D&cols=host%2Cservice%2C%40tags.n%2C%40tags.rnode&fromUser=true&messageDisplay=inline&refresh_mode=paused&storage=flex_tier&stream_sort=time%2Casc&top_n=10&top_o=top&viz=timeseries&x_missing=true&from_ts=1714716000000&to_ts=1714722300000&live=false.

(The approach I suggest will be easier if we can be sure that a single leaseholder will be returned from the trace, as per my above comment.)

Reply:
Yes, this makes sense. I'll look into doing this instead!